test/features/app/spec.js

```diff
@@ -188,84 +165,135 @@ function cmp(result, expected, errMsg) {
+        if (expBody.constructor === String) {
+            resBody = JSON.stringify(resBody);
         } else {
-            res.body = JSON.parse(res.body);
+            resBody = JSON.parse(resBody);
         }
     }
-    // check that the body type is the same
-    if (expRes.body.constructor !== res.body.constructor) {
-        throw new Error(
-            `Expected body type ${expRes.body.constructor} but got ${res.body.constructor}`
-        );
+    if (expBody.constructor === Object) {
+        Object.keys(expBody).forEach((key) => {
+            const val = expBody[key];
+             // eslint-disable-next-line
+            assert.deepEqual(resBody.hasOwnProperty(key), true, `Body field ${key} not found in response!`);
+            if (val.constructor === Object) {
+                validateBody(resBody[key], val);
+            } else if (val.constructor === Array) {
+                validateArray(val, resBody[key], key);
+            } else {
+                cmp(resBody[key], val, `${key} body field mismatch!`);
+            }
+        });
+    } else if (Array.isArray(expBody)) {
+        validateArray(expBody, resBody, 'body');
+    } else {
+        cmp(resBody, expBody, 'Body mismatch!');
     }
+    return true;
+}
 
-    // compare the bodies
-    cmp(res.body, expRes.body, 'Body mismatch!');
+function validateTestResponse(testCase, res) {
+    const expRes = testCase.response;
 
-    return true;
+    assert.deepEqual(res.status, expRes.status);
 
-}
+    if (expRes.headers && !res.headers) {return false; } // eslint-disable-line
 
+    Object.keys(expRes.headers).forEach((key) => {
+        const val = expRes.headers[key];
+        // eslint-disable-next-line
+        assert.deepEqual(res.headers.hasOwnProperty(key), true, `Header ${key} not found in response!`);
+        cmp(res.headers[key], val, `${key} header mismatch!`);
+    });
 
-describe('Swagger spec', function() {
+    validateBody(res.body || '', expRes.body);
+}
 
-    // the variable holding the spec
-    let spec = staticSpecLoad();
-    // default params, if given
-    let defParams = spec['x-default-params'] || {};
+describe('Swagger spec', function () {
 
-    this.timeout(20000); // eslint-disable-line no-invalid-this
+    this.timeout(20000);
 
-    before(() => {
-        return server.start();
-    });
+    before(() => server.start());
+    after(() => server.stop());
 
     it('get the spec', () => {
-        return preq.get(`${server.config.uri}?spec`)
+        baseUrl = server.config.uri;
+        return preq.get(`${baseUrl}?spec`)
         .then((res) => {
             assert.status(200);
             assert.contentType(res, 'application/json');
             assert.notDeepEqual(res.body, undefined, 'No body received!');
+            // save a copy
             spec = res.body;
+            return spec;
+        })
+        .then((spec) => {
+            const routeTests = () => {
+                before(() => server.start());
```

Comment:
Why are you starting/stopping for each test when we already have a service running?

Reply:
That's the workaround for the dynamic tests since they execute beyond the after() call. Some or all of the tests would return 504s since the service had shutdown.

Reply:
So why are we forcing them to be parallel then? What's the point?

Reply:
Let's either then find a proper solution (use a promise loop to check for running tests in `after()` or not use parallel at all.

Reply:
This is not start/stopping for each test, it's start-stopping for each generated test suite. 

Here's what's happening here.
1. mocha starts and executes synchronous code for all the test files. At this point no tests are executed, they're registered with mocha with appropriate before/after hook sequence
2. As we get to this set of tests, only static tests are executed. The dynamic tests are registered afterwards, but they apparently are appened to some internal array of tests in mocha. And they do not respect the before/after from static tests.
3. No matter whether we do that in parallel or sequentially, we need to start/stop a server for these tests.

This is how it's done in RESTBase too, so I think it's fine.